File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];

        result[0] = 1;
        for (int i = 1; i < n; i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }

        int prev = nums[n - 1];
        nums[n - 1] = 1;
        for (int i = n - 2; i >= 0; i--) {
            int temp = nums[i];
            nums[i] = nums[i + 1] * prev;
            prev = temp;
        }

        for (int i = 0; i < n; i++) {
            result[i] *= nums[i];
        }

        return result;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        // vector<int> prefixProduct(nums.size(), 1);
        // vector<int> suffixProduct(nums.size(), 1);
        // for (int i = 1; i < nums.size(); i++) {
        //     prefixProduct[i] = prefixProduct[i - 1] * nums[i - 1];
        // }
        // for (int i = nums.size() - 2; i >= 0; i--) {
        //     suffixProduct[i] = suffixProduct[i + 1] * nums[i + 1];
        // }

        // vector<int> result(nums.size(), 1);
        // for (int i = 0; i < nums.size(); i++) {
        //     result[i] = prefixProduct[i] * suffixProduct[i];
        // }

        // return result;

        vector<int> result(nums.size(), 1);
        for (int i = 1; i < nums.size(); i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }
        int R = 1;
        for (int i = nums.size() - 1; i >= 0; i--) {
            result[i] = result[i] * R;
            R *= nums[i];
        }

        return result;
    }
};